sql/core/src/main/scala/org/apache/spark/sql/sources/rules.scala

```diff
@@ -101,13 +101,17 @@ private[sql] case class PreWriteCheck(catalog: Catalog) extends (LogicalPlan =>
           }
         }
 
-      case i @ logical.InsertIntoTable(
-        l: LogicalRelation, partition, query, overwrite, ifNotExists)
-          if !l.isInstanceOf[InsertableRelation] =>
+      case logical.InsertIntoTable(LogicalRelation(_: InsertableRelation), _, _, _, _) =>
+        // OK
+
+      case logical.InsertIntoTable(LogicalRelation(_: FSBasedRelation), _, _, _, _) =>
+        // OK
+
+      case logical.InsertIntoTable(l: LogicalRelation, _, _, _, _) =>
```

Comment:
Maybe group these three lines closer together so it is obvious that they are a whitelist followed by a failure case.